File: aws_interface/resource/templates/python3/aws_interface.py
import requests
import json
import base64
import os


class Client:
    def __init__(self):
        self.session_id = None
        self.guest_id = None
        with open('manifest.json', 'r') as fp:
            self.manifest = json.load(fp)

    @property
    def _recipe_keys(self):
        return self.manifest['recipe_keys']

    def _get_recipe_manifest(self, recipe_key):
        if recipe_key not in self._recipe_keys:
            raise Exception('recipe_key must be in {}'.format(self._recipe_keys))
        else:
            return self.manifest[recipe_key]

    def _get_api_list(self, recipe_key):
        manifest = self._get_recipe_manifest(recipe_key)
        return manifest['cloud_apis']

    def _get_api_url(self, recipe_key):
        manifest = self._get_recipe_manifest(recipe_key)
        return manifest['rest_api_url']

    def _call_api(self, recipe_key, api_name, data=None):
        manifest = self._get_recipe_manifest(recipe_key)
        if not data:
            data = {}
        url = self._get_api_url(recipe_key)
        data['cloud_api_name'] = api_name
        data['recipe_key'] = recipe_key
        if self.session_id:
            data['session_id'] = self.session_id
        data = json.dumps(data)
        resp = _post(url, data)
        return resp.json().get('body', {'error': '404', 'message': 'NO RESPONSE'})

    def _auth(self, api_name, data):
        return self._call_api('auth', api_name, data)

    def _database(self, api_name, data):
        return self._call_api('database', api_name, data)

    def _storage(self, api_name, data):
        return self._call_api('storage', api_name, data)

    def auth_register(self, email, password, extra={}):
        response = self._auth('register', {
            'email': email,
            'password': password,
            'extra': extra
        })
        return response

    def auth_login(self, email, password):
        response = self._auth('login', {
            'email': email,
            'password': password
        })
        self.session_id = response.get('session_id', None)
        return response

    def auth_get_user(self, user_id):
        response = self._auth('get_user', {
            'user_id': user_id
        })
        return response

    def auth_get_users(self, start_key=None):
        response = self._auth('get_users', {
            'start_key': start_key,
        })
        return response

    def auth_logout(self):
        response = self._auth('logout', {
            'session_id': self.session_id
        })
        self.session_id = None
        return response

    def auth_guest(self, guest_id=None):
        data = {}
        if guest_id:
            data['guest_id'] = guest_id
        response = self._auth('guest', data)
        self.session_id = response.get('session_id', None)
        return response

    # def auth_delete_user(self):

    def database_create_item(self, item, partition, read_groups, write_groups):
        response = self._database('create_item', {
            'item': item,
            'partition': partition,
            'read_groups': read_groups,
            'write_groups': write_groups,
        })
        return response

    def database_delete_item(self, item_id):
        response = self._database('delete_item', {
            'item_id': item_id
        })
        return response

    def database_get_item(self, item_id):
        response = self._database('get_item', {
            'item_id': item_id
        })
        return response

    def database_get_items(self, partition, start_key=None, limit=None):
        response = self._database('get_items', {
            'partition': partition,
            'start_key': start_key,
            'limit': limit,
        })
        return response

    def database_put_item_field(self, item_id, field_name, field_value):
        response = self._database('put_item_field', {
            'item_id': item_id,
            'field_name': field_name,
            'field_value': field_value,
        })
        return response

    def database_update_item(self, item_id, item, read_groups, write_groups):
        response = self._database('update_item', {
            'item_id': item_id,
            'item': item,
            'read_groups': read_groups,
            'write_groups': write_groups,
        })
        return response

    def database_query_items(self, partition, query, start_key=None, limit=None):
        response = self._database('query_items', {
            'partition': partition,
            'query': query,
            'start_key': start_key,
            'limit': limit,
        })
        return response

    def _storage_delete_b64(self, file_id):
        response = self._storage('delete_b64', {
            'file_id': file_id
        })
        return response

    def _storage_download_b64_chunk(self, file_id):
        response = self._storage('download_b64', {
            'file_id': file_id
        })
        return response

    def _storage_upload_b64_chunk(self, parent_file_id, file_name, file_b64, read_groups, write_groups):
        response = self._storage('upload_b64', {
            'parent_file_id': parent_file_id,
            'file_name': file_name,
            'file_b64': file_b64,
            'read_groups': read_groups,
            'write_groups': write_groups,
        })
        return response

    def storage_delete_file(self, file_id):
        self._storage_delete_b64(file_id)

    def storage_download_file(self, file_id, download_path):
        """
        :param file_id: file_id that is uploaded on aws via storage_upload_file
        :param download_path: str
        :return:
        """
        string_file_b64 = None
        file_name = 'file'
        while file_id:
            result = self._storage_download_b64_chunk(file_id)
            file_id = result.get('parent_file_id', None)
            file_name = result.get('file_name', file_name)
            if string_file_b64:
                string_file_b64 = result.get('file_b64') + string_file_b64
            else:
                string_file_b64 = result.get('file_b64')

        string_b64 = string_file_b64.encode('utf-8')
        file_bin = base64.b64decode(string_b64)

        with open(download_path, 'wb+') as file_obj:
            file_obj.seek(0)
            file_obj.write(file_bin)

    def storage_upload_file(self, file_path, read_groups=['owner', 'user'], write_groups=['owner']):
        def divide_chunks(text, n):
            for i in range(0, len(text), n):
                yield text[i:i + n]

        file_name = os.path.basename(file_path)
        with open(file_path, 'rb') as file_obj:
            raw_base64 = file_obj.read()

        raw_base64 = base64.b64encode(raw_base64)
        raw_base64 = raw_base64.decode('utf-8')

        base64_chunks = divide_chunks(raw_base64, 1024 * 1024 * 6)  # 4mb
        parent_file_id = None
        for base64_chunk in base64_chunks:
            result = self._storage_upload_b64_chunk(parent_file_id, file_name, base64_chunk, read_groups, write_groups)
            parent_file_id = result.get('file_id')
        return result


def _post(url, data):
    response = requests.post(url, data)
    return response


def example():
    email = 'email@example.com'
    password = 'password'
    client = Client()

    response = client.auth_register(email, password)
    print('auth_register response: {}'.format(response))

    response = client.auth_login(email, password)
    print('auth_login response: {}'.format(response))

    response = client.database_create_item({
        'type': 'test',
    }, 'test', ['owner'], ['owner'])
    print('database_create_item response: {}'.format(response))

    response = client.storage_upload_file('manifest.json')
    print(response)

    client.storage_download_file(response['file_id'], 'download')


if __name__ == '__main__':  # SHOW EXAMPLE
    example()

File: aws_interface/resource/templates/nodejs/aws_interface.js
const request = require('request');
const base_url = '{{REST_API_URL}}';

class Client{

    constructor() {
        this.session_id = null;
    }

    static _get(object, key, default_value=null) {
        var result = object[key];
        return (typeof result !== "undefined") ? result : default_value;
    }

    _call_api(service_type, function_name, data, callback) {
        if (data == null){
            data = {};
        }
        data['module_name'] = 'cloud.' + service_type + '.' + function_name;
        if (this.session_id != null){
            data['session_id'] = this.session_id;
        }
        this._post(base_url, data, callback);
    }

    _post(url, data, callback){
        let options = {
            uri: base_url,
            method: 'POST',
            json: data,
        };
        request(options, function (error, response, body) {
            if (error) {
                callback({
                    'error': error,
                    'message': error.message,
                });
            }else{
                callback(body['body']);
            }
        });
    }

    _auth(api_name, data, callback) {
        let self = this;
        this._call_api('auth', api_name, data, function (data) {
            self.log_create_log('auth', api_name, null, function (data) {
                
            });
            callback(data);
        });
    }

    _database(api_name, data, callback) {
        let self = this;
        this._call_api('database', api_name, data, function (data) {
            self.log_create_log('database', api_name, null, function (data) {

            });
            callback(data);
        });
    }

    _storage(api_name, data, callback) {
        let self = this;
        this._call_api('storage', api_name, data, function (data) {
            self.log_create_log('storage', api_name, null, function (data) {

            });
            callback(data);
        });
    }

    _log(api_name, data, callback) {
        this._call_api('log', api_name, data, callback);
    }

    auth_register(email, password, extra={}, callback) {
        this._auth('register', {
            'email': email,
            'password': password,
            'extra': extra,
        }, callback);
    }

    auth_login(email, password, callback) {
        let self = this;
        this._auth('login', {
            'email': email,
            'password': password,
        }, function (data) {
            if ('session_id' in data){
                self.session_id = Client._get(data,'session_id');
            }
            callback(data);
        });
    }

    auth_get_user(user_id, callback) {
        this._auth('get_user', {
            'user_id': user_id,
        }, callback);
    }

    auth_get_users(start_key=null, callback) {
        this._auth('get_users', {
            'start_key': start_key,
        }, callback);
    }

    auth_logout(callback) {
        this._auth('logout', {
            'session_id': this.session_id,
        }, callback);
    }

    auth_guest(guest_id=null, callback) {
        let self = this;
        this._auth('guest', {
            'guest_id': guest_id,
        }, function (data) {
            self.session_id = Client._get(data,'session_id');
            callback(data);
        });
    }

    database_create_item(item, partition, read_groups, write_groups, callback) {
        this._database('create_item', {
            'item': item,
            'partition': partition,
            'read_groups': read_groups,
            'write_groups': write_groups,
        }, callback);
    }

    database_delete_item(item_id, callback) {
        this._database('delete_item', {
            'item_id': item_id,
        }, callback);
    }

    database_get_item(item_id, callback) {
        this._database('get_item', {
            'item_id': item_id,
        }, callback);
    }

    database_get_items(partition, start_key=null, limit=100, callback) {
        this._database('get_items', {
            'partition': partition,
            'start_key': start_key,
            'limit': limit,
        }, callback);
    }

    database_put_item_field(item_id, field_name, field_value, callback) {
        this._database('put_item_field', {
            'item_id': item_id,
            'field_name': field_name,
            'field_value': field_value,
        }, callback);
    }

    database_update_item(item_id, item, read_groups, write_groups, callback) {
        this._database('update_item', {
            'item_id': item_id,
            'item': item,
            'read_groups': read_groups,
            'write_groups': write_groups,
        }, callback);
    }

    database_query_items(partition, query, start_key=null, limit=100, reverse=false, callback) {
        this._database('query_items', {
            'partition': partition,
            'query': query,
            'start_key': start_key,
            'limit': limit,
            'reverse': reverse,
        }, callback);
    }

    storage_delete_b64(file_id, callback) {
        this._storage('delete_b64', {
            'file_id': file_id,
        }, callback);
    }

    storage_download_b64_chunk(file_id, callback) {
        this._storage('download_b64', {
            'file_id': file_id,
        }, callback);
    }

    storage_upload_b64_chunk(parent_file_id, file_name, file_b64, read_groups, write_groups, callback) {
        this._storage('upload_b64', {
            'parent_file_id': parent_file_id,
            'file_name': file_name,
            'file_b64': file_b64,
            'read_groups': read_groups,
            'write_groups': write_groups,
        }, callback);
    }

    storage_delete_file(file_id, callback) {
        this.storage_delete_b64(file_id, callback);
    }

    storage_download_file(file_id, callback_file) {
        let self = this;
        var string_file_b64 = null;
        var file_name = 'file';
        function download(file_id, callback){
            self.storage_download_b64_chunk(file_id, function (result) {
                file_id = Client._get(result, 'parent_file_id', null);
                file_name = Client._get(result,'file_name', file_name);
                if (string_file_b64 != null){
                    string_file_b64 = Client._get(result, 'file_b64') + string_file_b64;
                }else{
                    string_file_b64 = Client._get(result, 'file_b64');
                }
                if (file_id == null){
                    callback(string_file_b64, result);
                }else{
                    download(file_id, callback);
                }
            });
        }
        download(file_id, function (string_file_b64, result) {
            if (string_file_b64 == null){
                callback_file(null);
                console.error(result);
            }else{
                let file_bin = Buffer.from(string_file_b64, 'base64');
                callback_file(file_bin);
            }
        });
    }

    storage_upload_file(bin, read_groups, write_groups, callback) {
        let self = this;
        function *div_chunks(text, n){
            for (var i = 0; i < text.length; i+= n){
                yield text.slice(i, i + n);
            }
        }
        var buff = Buffer.from(bin, 'utf8');
        let base64_data = buff.toString('base64');
        var raw_base64 = Buffer.from(base64_data, 'utf8');
        var base64_chunks = div_chunks(raw_base64, 1024 * 1024 * 6);
        var parent_file_id = null;
        let file_name = "file";
        function upload(parent_file_id, base64_chunk, callback){
            base64_chunk = base64_chunk.toString();
            self.storage_upload_b64_chunk(parent_file_id, file_name, base64_chunk, read_groups, write_groups, function (data) {
                var parent_file_id = data['file_id'];
                var next_base64_chunk = base64_chunks.next();
                if (next_base64_chunk.done){
                    callback(data);
                }else{
                    upload(parent_file_id, next_base64_chunk.value, callback);
                }
            });
        }
        upload(parent_file_id, base64_chunks.next().value, callback);
    }

    log_create_log(event_source, event_name, event_param, callback) {
        this._log('create_log', {
            'event_source': event_source,
            'event_name': event_name,
            'event_param': event_param,
        }, callback);
    }

}

module.exports.Client = Client;